feat: 캘린더 멤버 초대 링크 생성 및 CustomAuthenticationEntryPoint 적용

jooyeon0907 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 초대 링크 생성 기능이 구현되지 않음
- 토큰 인증 실패 시 403 에러만 반환 (에러 메세지 없음)

**TO-BE**
- Redis를 활용하여 초대 링크 관리
    - Redis 라이브러리 추가
    - application.yml 에 Redis 환경 설정
    - 캘린더 초대 링크 생성
        - UUID 값을 생성
        - redis에 UUID를 키(Key)로, 초대하는 캘린더의 ID 는 값(Value)으로 저장 (유효기간 24시간 설정)
        - 초대 링크에 UUID 값을 포함하여 반환
    - 캘린더 초대 링크를 통한 멤버 추가
        - 링크에서 UUID 값 추출
        - Redis 에 해당 UUID 키가 존재하는지 확인
            - 존재하지 않으면 -> "초대 링크가 만료되었거나 존재하지 않습니다."  응답
        - 초대 코드 사용 후 재사용 방지를 위해 Redis 에서 삭제
        - UserCalendar 엔티티에 초대된 사용자 추가
            - user_id , calendar_id 값을 저장

- CustomAuthenticationEntryPoint 적용
    - 토큰 인증 실패 시 403 에러 대신 401 에러 메시지 반환
    - 기존에 토큰 인증 실패 시 403 에러만 반환되고, 별다른 에러 메세지가 없었음.
    - 이를 개선하여 에러 메시지를 포함하고, 기존 응답 형식과 일치하도록 처리

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] Redis 연결 테스트 코드
    - @BeforeEach 어노테이션을 사용하여 테스트 실행 전에 Redis 에 초기 데이터를 설정
        - 'testKey' Key 에 'Hello, Redis!' 값을 설정
    - Redis 에 'testKey' 에 저장된 값을 읽어와서 assertEquals 를 사용하여 값이 예상과 일치하는지 확인
    - Redis 에 'testKey' 삭제 후 해당 키 값을 조회하여 null 이 반환되는지 확인
- [x] API 테스트 (Postman)
    - 캘린더 초대 링크 생성
        - JWT 토큰 인증 확인
        - Redis에 데이터가 추가된 것을 확인
        - 예외 처리
            - 토큰 인증 실패: "로그인 정보가 유효하지 않습니다."
            - 캘린더 멤버가 아님: "해당 사용자는 이 캘린더의 멤버가 아닙니다."
            - 캘린더 타입이 개인용: "개인용 캘린더는 멤버를 초대할 수 없습니다."
        - 초대 링크가 반환 성공 여부 확인
    - 초대 링크를 통한 멤버 추가
        - JWT 토큰 인증 확인
        - Redis에 데이터가 삭제된 것을 확인
        - 예외 처리
            - Redis에 키 값이 존재하지 않음: "초대 링크가 만료되었거나 존재하지 않습니다."
            - 존재하지 않는 캘린더: "캘린더 정보가 존재하지 않습니다."
            - 이미 초대된 멤버: "해당 캘린더에 이미 초대된 멤버입니다."
        - 캘린더 멤버 추가 성공 여부 확인

jooyeon0907 committed:
feat: 캘린더 멤버 초대 링크 생성 및 CustomAuthenticationEntryPoint 적용

jooyeon0907 committed:
feat: 캘린더 멤버 초대 링크 생성 및 CustomAuthenticationEntryPoint 적용

jooyeon0907 committed:
fix: 코드 수정

mingming-mentor commented on src/main/java/com/zerobase/timate/security/CustomAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,42 @@
+@Component
+public class CustomAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+	private final ObjectMapper objectMapper = new ObjectMapper();
```
기본적으로 Spring에서 만들어서 관리하는 objectMapper bean이 있기 때문에 해당 bean을 주입받으셔도 좋을 것 같아요

mingming-mentor commented on src/main/java/com/zerobase/timate/security/CustomAuthenticationEntryPoint.java:
```diff
@@ -0,0 +1,42 @@
+		response.getWriter().write(objectMapper.writeValueAsString(apiResponse));
+
+		// 상태 코드와 메시지를 로그로 출력 (디버깅용)
+		log.info("Response Status: " + response.getStatus());
```
+가 아니라 {} 를 쓰시면 좋겠습니다. 그리고 debug용이면 debug loglevel을 써도 좋겠습니다
```suggestion
		log.debug("Response Status: {}", response.getStatus());
```

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarMemberService.java:
```diff
@@ -136,5 +147,58 @@ private void changeMaster(Long userId, Long newMasterId, Long calendarId) {
+	public String generateInviteLink(Long userId, Long calendarId) {
+		// 해당 사용자가 캘린더의 멤버인지 확인
+		UserCalendar userCalendar = commonService.getUserCalendar(userId, calendarId);
+		if (userCalendar.getCalendar().getType().equals(CalendarType.PRIVATE)) {
```
enum은 == 로 비교하시면 좋겠습니다

jooyeon0907 replied:
직전 pr 리뷰와 같은 내용인데
해당 코드가 이전에 작성했던 거여서 미쳐 확인을 못했네요 ㅜㅜ 다음에 pr 올리기 전에 한 번 더 확인해 보겠습니다

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarMemberService.java:
```diff
@@ -136,5 +147,58 @@ private void changeMaster(Long userId, Long newMasterId, Long calendarId) {
+        String redisKey = getInvitationRedisKey(calendarUuid);
+
+        // Redis에 초대 코드 저장 (유효기간 24시간)
+        redisTemplate.opsForValue().set(redisKey, calendarId.toString(), Duration.ofHours(24));
```
calendarId를 String으로 변환할 필요 없이 RedisTemplate bean을 만들때 value를 Long으로 받도록 설정하시면 좋겠네요.
불필요한 형변환을 줄일 수 있을 것 같습니다

jooyeon0907 replied:
네 피드백 반영하여 머지하겠습니다! 감사합니다

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarMemberService.java:
```diff
@@ -136,5 +147,58 @@ private void changeMaster(Long userId, Long newMasterId, Long calendarId) {
+	 * 초대 링크 검증
+	 */
+	public Long validateInvitation(String inviteCode) {
+		String redisKey = getInvitationRedisKey(inviteCode);
+		String calendarId = redisTemplate.opsForValue().get(redisKey);
```
```suggestion
		String calendarId = redisTemplate.opsForValue().get(getInvitationRedisKey(inviteCode));
```
redisKey가 여려번 코드에서 사용되는 것이 아니라 한번만 사용된다면 별도의 변수로 선언하지 않는 것이 가독성이나 리소스측면에서 좋습니다

mingming-mentor reviewed (approved):
수고 많으셨습니다

jooyeon0907 committed:
fix: 코드 수정
- redis에 저장되는 calendarId를 불필요한 형변환을 하지 않고(Long -> String) Long 타입 그대로 저장하기 위해 수정

jooyeon0907 committed:
Merge remote-tracking branch 'origin/feature/calendar_invitation_link' into feature/calendar_invitation_link

# Conflicts:
#	src/main/java/com/zerobase/timate/security/CustomAuthenticationEntryPoint.java
#	src/main/java/com/zerobase/timate/service/CalendarMemberService.java
#	src/test/java/com/zerobase/timate/redis/RedisConnectionTest.java
